### ISSUE TITLE
[dev.icinga.com #2955] libicinga: avoid multiple alloc/free calls in kvvec; use assignment in buf2kvvec(); s/kvvec_grow/kvvec_resize/

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icinga-core#1070]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/2955

**Created by mfriedrich on 2012-08-05 12:30:42 +00:00**

Assignee: _mfriedrich_
Status: _Closed_  (closed on _2012-08-22 18:19:30 +00:00_)
Target Version: _exp_
Last Update: _2012-08-22 18:19:30 +00:00 (in Redmine)_

---

    lib/kvvec: Add possibility to avoid multiple alloc()/free() calls
    It's quite simple to allow callers to provide a pointer to their
    own previously allocated key/value vector, and this allows them to
    avoid a large amount of needless memory munging, so it proves to be
    quite a useful performance improvement as well.

    Signed-off-by: Andreas Ericsson 

    lib/kvvec: Use assignment rather than copying in buf2kvvec()
    This saves an enormous amount of work, since we no longer have to
    get fresh memory for each check result reported in by the workers.
    Since new memory would inevitably be located in a different page
    than the I/O cache buffer we've got allocated for the worker, that
    means we can now avoid at least one cache-line fill and a (relatively)
    huge constant overhead of check result reception.

    Signed-off-by: Andreas Ericsson 

    lib/kvvec: s/kvvec_grow/kvvec_resize/
    This bug stems from the time when 'grow' actually meant 'resize',
    but since the 'buf2kvvec_prealloc()' introduction it could (and did)
    mean that callers might get key/value vectors that grew out of
    control, by roughly 22 indices per new check-result caught. Not too
    hot in long-running daemons.

    Signed-off-by: Andreas Ericsson 

**Changesets**

_2012-08-05 13:00:40 +00:00_ by _mfriedrich_ 0fe097c440c267eca9a33526eca8a6089629b317

    libicinga: add possibility to avoid multiple allo()/free() calls #2955
    
    it's rather simple to allow callers to provide a pointer to their own
    previously allocated key/value vector, and this allows them to avoid a
    large amount of needless memory munging, so it proves to be quite useful
    in performance improvements too.
    
    refs #2955

_2012-08-05 13:27:46 +00:00_ by _mfriedrich_ ed894e8bbd4894c74ca2f79341f9f420bb61c595

    libicinga: use assignment rather than copying in buf2kvvec() #2955
    
    this adds the flags directive to buf2kvvec() calls, allowing to set to
    copy or append all data.
    
    basically, we don't need to get fresh memory for each checkresults
    reported in by the workers. since new memory would inevitable be located
    in a different page than the I/O cache buffer we've got allocated for
    the worker, that means we can now avoid at least one cache-line fill and
    a (relatively) huge constant overhead of check result reception.
    
    refs #2955

_2012-08-05 21:19:57 +00:00_ by _mfriedrich_ 5f7eb89451e37076f412c8574d8417a7132e5698

    core: fix logical bug in handling log returned results first
    
    refs #2961
    refs #2955

_2012-08-05 21:31:19 +00:00_ by _mfriedrich_ 8bfadf52654c29b28f338d261bec70e06bc6f052

    libicinga: fix faulty brackets in kvvec; possible segfault on worker_result_handler [WIP]
    
    refs #2961
    refs #2955
